### DISCUSSION TITLE
How to run multiple workers in container?

### DISCUSSION THREAD

**Discussion by jonashaag** [trinodb/trino#20020] [Q&A]
I've started 1 coordinator + 1 worker, each in a Podman container. They have different Node IDs, but when I click on the worker UI (`worker.html?<node ID>`) I'm shown the info of the coordinator. The worker is also never used.

Both are show as having the same IP address; maybe that's why the UI confused them? They both are shown as having IP 10.0.2.100 (which is NOT their actual IP, but a special IP by Podman).

<img width="1286" alt="Screenshot 2023-12-05 at 09 02 34" src="https://github.com/trinodb/trino/assets/175722/450822f1-565d-456a-85ec-049bc1f024b6">
<img width="1201" alt="Screenshot 2023-12-05 at 09 02 08" src="https://github.com/trinodb/trino/assets/175722/1086a4a0-35c3-4dec-8a57-4819ca58bbfb">

**Comment by mosabua** [comment]
Looks like they are completely isolated from each other network wise. How did you deploy them?

**Comment by jonashaag** [comment]
They are not. The worker can connect to the coordinator (but not receive any work). I can ping/telnet between the machines/servers using the correct IPs and ports. But Trino's mechanism of detecting the IP (and maybe port?) seems to be not working here.

Deployment: Manually start two barebones Ubuntu containers and run the coordinator/worker instance using the launcher script.

**Comment by mosabua** [comment]
There is no detection mechanism.. you need to configure the discovery server URL (and port) in each node and have that set in config.properties 

When you go to the web ui on the coordinator .. does it show the correct number of workers?

Also btw.. the helm chart sorts this all out for you on k8s

**Reply by jonashaag** [reply]
Yes. It shows the correct number but the UI confuses one for the other. I'll post a screenshot to make it more clear.

**Reply by jonashaag** [reply]
I added screenshots. Note how the UI confuses the two workers.

**Comment by mosabua** [comment]
Somehow your podman setup messes with the IP numbers of the workers ..

**Reply by jonashaag** [reply]
Yep :(

**Reply by aradovic-sym** [reply]
Ever solved this?

**Reply by aradovic-sym** [reply]
For others, I have solved it.

I'm using Trino inside of Nomad w/ the Podman plugin. Essentially, you need to provide a `node.properties` with the appropriate settigns to set the internal and external IP. I'll share how I do it w/ Nomad, you figure it out w/ Podman alone (change the entrypoint w/ a script that templates the file or sth.).

Code:

```hcl
      template {
        data = <<-EOH
          node.environment=[[ .environment ]]
          node.data-dir=/data/trino
          # These are undocumented settings from Trino's framework, Airlift,
          # built by the same team (practically 1/2 of the Trino codebase,
          # confusingly).
          #
          # See https://github.com/airlift/airlift/blob/master/node/src/main/java/io/airlift/node/NodeConfig.java
          # found via https://github.com/trinodb/trino/discussions/18584
          #
          # They are required so that the workers and coordinators announce an
          # address to the discovery servicet that's actually routable within
          # Nomad, because the default IP is not (and is the same everywhere).
          #
          # WARNING: this is an undocumented feature and might break in the
          # future
          node.internal-address={{ env "NOMAD_IP_trino_coordinator_http" }}
          node.external-address={{ env "NOMAD_IP_trino_coordinator_http" }}
        EOH
        destination = "local/etc/trino/node.properties"
      }
```

Note that you have to provide `node.environment`, and possibly `node.data-dir` if you're replacing the `node.properties` because those are in the original and Trino won't work without them.

**Reply by jennevv** [reply]
This solution worked for me.
Thank you @aradovic-at-symmetry!

**Comment by hashhar** [comment]
cc: @electrum might know off the top of his head.
